docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,106 @@
+
+We have [prototyped](https://github.com/fermyon/spin/pull/2229) dispatching these requests directly to the component handling the route, but this was in the context of an implementation detail of Spin.  The requirement for this project is for an application to declare that it _requires_ dispatch to run locally on the node, and for hosts that cannot meet that prerequisite to reject the application.  This allows developers to write applications as a network of chained microservices, without having to be concerned about what happens if the microservices are not co-located.
+
+# User experience
+
+## Expressing a requirement for chaining
+
+To express that the application requires chaining for self-requests, add the following to `spin.toml`:
+
+```toml
+[application.prerequisites]
+service_chaining = "required"
+```
+
+**BIKESHED:** Or something more explicit such as `direct_self_requests`.  Or `prerequisites` could be an array of capabilities that the host must provide (`prerequisites = ["service-chaining"]`).  Also, we _could_ call this something like `environment` or `host` instead, but I'm kind of inclined to leave it generic as to what prerequisites might go in here.
+
+Entries in `prerequisites` are considered must-understand.  For the section itself, that has to happen at a `serde` level since we can't retrofit it to existing builds.  But `serde` errors are not very lovely, and I'd prefer hosts in future to be able to say "hey I don't support your `gpu` requirement" rather than "unknown field serialisation error."
+
+## Requesting a component via chaining
+
+There are two ways we could express a HTTP request to a chained component:
+
+1. Use the existing self-request style.  The host _may_ chain self-requests, and _must_ chain them if the "require chaining" setting is on.  This is super easy for the application developer, at both the send and receive end.  However, it becomes more complex for the host, as the host now has to route the request - this drags in much more of the HTTP trigger plumbing.  (The prototype did this and it was egregiously hairy.)
+
+2. Address requests directly to the destination components by name. We still need a URL, so this would introduce a special host name such as `spin.internal`.  Then the URL `http://spin.internal/accounts` would invoke the `accounts` component.  This bypasses routing, which should simplify implementation.  It also opens the door to private endpoints, where an internal microservice is not exposed to the network (does not have a HTTP route) and so is accessible only from within the application.  However, chained components which parse URLs may need to take extra care to correctly handle both routed and chaining URLs.
```

Comment:
> chained components which parse URLs may need to take extra care to correctly handle both routed and chaining URLs.

This may be an argument for a slightly different URL scheme: `http://<component>.spin.internal/<any path allowed here>`

Reply:
If this is in reference to URL parsing, it doesn't in itself solve the problem: components will still need to be aware that they could be called via two different paths.  But I do like that it keeps the component name safely separated from the rest of the path, so it gets discarded "naturally" without the receiver needing to strip segments.  As you note elsewhere, it also allows more granular restriction based on hosts, so I'm in favour.